backend/controller/evaljob_controller.go

```diff
@@ -0,0 +1,568 @@
+		if err := r.deleteJobPod(ctx, job); err != nil {
+			// leave the state as is and retry again
+			log.Error(err, "failed to delete pod. scheduled a retry after 10 seconds")
+			return ctrl.Result{Requeue: true, RequeueAfter: time.Second * 10}, err
+		}
+	}
+
+	err := r.Status().Update(ctx, job)
+	if err != nil {
+		log.Error(err, "failed to update status for cancellation")
+	}
+	r.Recorder.Event(job, "Normal", "Cancelled",
+		fmt.Sprintf("Tthe EvalJob %s in namespace %s has cancelled and changed its state to Complete",
+			job.Name,
+			job.Namespace))
+	return ctrl.Result{}, err
+}
+
+func (r *EvalJobReconciler) createPod(job *lmevalservicev1beta1.EvalJob) *corev1.Pod {
+	var allowPrivilegeEscalation = false
+	var runAsNonRootUser = true
+	var ownerRefController = true
+	var runAsUser int64 = 1001030000
+
+	// Then compose the Pod CR
+	pod := corev1.Pod{
+		TypeMeta: v1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name:      job.Name,
+			Namespace: job.Namespace,
+			OwnerReferences: []v1.OwnerReference{
+				{
+					APIVersion: job.APIVersion,
+					Kind:       job.Kind,
+					Name:       job.Name,
+					Controller: &ownerRefController,
+					UID:        job.UID,
+				},
+			},
+			Labels: map[string]string{
+				"app.kubernetes.io/name": "fms-lm-eval-service",
+			},
+		},
+		Spec: corev1.PodSpec{
+			InitContainers: []corev1.Container{
+				{
+					Name:            "driver",
+					Image:           r.options.DriverImage,
+					ImagePullPolicy: corev1.PullAlways,
+					Command:         []string{DriverPath, "--copy", DestDriverPath},
+					SecurityContext: &corev1.SecurityContext{
+						AllowPrivilegeEscalation: &allowPrivilegeEscalation,
+						RunAsUser:                &runAsUser,
+						Capabilities: &corev1.Capabilities{
+							Drop: []corev1.Capability{
+								"ALL",
+							},
+						},
+					},
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							Name:      "shared",
+							MountPath: "/opt/app-root/src/bin",
+						},
+					},
+				},
+			},
+			Containers: []corev1.Container{
+				{
+					Name:            "main",
+					Image:           r.options.PodImage,
+					ImagePullPolicy: corev1.PullAlways,
+					Env: []corev1.EnvVar{
+						{
+							Name: "GENAI_KEY",
+							ValueFrom: &corev1.EnvVarSource{
+								SecretKeyRef: &corev1.SecretKeySelector{
+									Key: "key",
+									LocalObjectReference: corev1.LocalObjectReference{
+										Name: "genai-key",
+									},
+								},
+							},
+						},
+					},
+					Command: generateCmd(job),
+					Args:    generateArgs(job),
+					SecurityContext: &corev1.SecurityContext{
+						AllowPrivilegeEscalation: &allowPrivilegeEscalation,
+						RunAsUser:                &runAsUser,
+						Capabilities: &corev1.Capabilities{
+							Drop: []corev1.Capability{
+								"ALL",
+							},
+						},
+					},
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							Name:      "shared",
+							MountPath: "/opt/app-root/src/bin",
```

Comment:
Interesting. Why is this emptydir mounting to a `bin` path?

Reply:
oooh, I think I see now. Is this getting the binaries dropped into it during the init container? Why not just require that the pod's image have the necessary binaries?

Reply:
Ah! I think I understand the role of the `driver` now. Is this basically just a wrapper around executing the python script that will push the status updates to the parent resource?

Reply:
Assuming this is correct, I think we may need to be pretty careful here using a push model for status updates rather than a pull model where the controller polls the pod. The reason for this is that some security-focused users will want to isolate the `pod` where the job body runs in a namespace that fully disallows any RBAC for pods.

Reply:
that's used to copy the driver from the `driver` image in the init container to `lm-eval + unitxt` image in the main container. So in the main container, we are still running the `driver` and have the driver to run the "lm-eval + unitxt" python program.